### PR TITLE
make env vars reviewable

### PR DESCRIPTION
**Pull request by grosser**
problem: when a user enters per-deploy environment variables they did not show up in buddy review window (neither did stage var changes)

@zendesk/compute @zendesk/bre 

Screen Shot 2019-04-04 at 4 02 41 PM

also:
 - rework priority to be additive instead of callback based
 - make deploy review and confirm consistent (not removing trailing space)
 - allow any params in deploy/new
 - sort env vars correctly by deploy/stage/project/env/deploy-group

### Risks
 - medium: project-wide env vars are now directly in the job executors environment which could lead to surprises
 - low: previously all deploy/stage env vars were injected into all deploys no matter if they matched the scope

### CODE REVIEW

File: plugins/env/app/controllers/environment_variable_groups_controller.rb [hunk 1]
```diff
@@ -54,7 +54,7 @@ def preview
       @project = Project.find(params[:project_id])
     end
 
-    @groups = SamsonEnv.env_groups(@project, deploy_groups, preview: true)
+    @groups = SamsonEnv.env_groups(Deploy.new(project: @project), deploy_groups, preview: true)
```
Comment (adammw) [inline plugins/env/app/controllers/environment_variable_groups_controller.rb]: is deploy_groups not part of the `Deploy` object?

**Comment by grosser** [comment]
no

File: plugins/aws_sts/lib/samson_aws_sts/samson_plugin.rb [hunk 2]
```diff
@@ -72,8 +72,8 @@ def session_name(deploy)
   ]
 end
 
-Samson::Hooks.callback :deploy_env do |deploy|
-  next {} if deploy.stage.aws_sts_iam_role_arn.blank?
+Samson::Hooks.callback :deploy_env do |deploy, *|
```
Comment (adammw) [inline plugins/aws_sts/lib/samson_aws_sts/samson_plugin.rb]: _deploy_group would be clearer what the second variable you're expecting is...

File: lib/samson/hooks.rb [hunk 3]
```diff
@@ -42,7 +42,6 @@ class UserError < StandardError
       :build_permitted_params,
       :buildkite_release_params,
       :can,
-      :deploy_group_env,
       :deploy_group_includes,
```
Comment (adammw) [inline lib/samson/hooks.rb]: do you document the signature of each hook somewhere? would be good to say what you expect to call deploy_env with... (deploy + deploy groups)

File: plugins/kubernetes/app/models/kubernetes/template_filler.rb [hunk 4]
```diff
@@ -468,8 +468,8 @@ def static_env
         env[:BLUE_GREEN] = blue_green_color if blue_green_color
 
         # env from plugins
-        plugin_envs = Samson::Hooks.fire(:deploy_group_env, project, @doc.deploy_group, resolve_secrets: false)
-        plugin_envs += Samson::Hooks.fire(:deploy_env, @doc.kubernetes_release.deploy) if @doc.kubernetes_release.deploy
+        deploy = @doc.kubernetes_release.deploy || Deploy.new(project: project)
+        plugin_envs = Samson::Hooks.fire(:deploy_env, deploy, @doc.deploy_group, resolve_secrets: false)
```
Comment (adammw) [inline plugins/kubernetes/app/models/kubernetes/template_filler.rb]: do you still need to pass through a separate deploy_group? can't you get that from the deploy?

File: lib/samson/hooks.rb (hunk 3 above)
Comment (grosser) [reply lib/samson/hooks.rb]: yeah, for now it's "look at the source" :)

File: plugins/kubernetes/app/models/kubernetes/template_filler.rb (hunk 4 above)
Comment (grosser) [reply plugins/kubernetes/app/models/kubernetes/template_filler.rb]: deploys don't have a deploy-group, they only know their stage which has multiple

File: plugins/env/app/models/environment_variable.rb [hunk 5]
```diff
@@ -92,13 +94,13 @@ def resolve_secrets(project, deploy_group, env, preview:)
     end
   end
 
-  # used by `priority` from GroupScope
-  def project?
-    parent_type == "Project"
+  def priority
+    [PARENT_PRIORITY.index(parent_type) || 999, super]
```
Comment (adammw) [inline plugins/env/app/models/environment_variable.rb]: why use a lookup here when you used a case statement in another class?
also didn't you get rid of the array?

File: plugins/aws_sts/lib/samson_aws_sts/samson_plugin.rb (hunk 2 above)
Comment (grosser) [reply plugins/aws_sts/lib/samson_aws_sts/samson_plugin.rb]: added

File: plugins/env/app/models/environment_variable.rb (hunk 5 above)
Comment (grosser) [reply plugins/env/app/models/environment_variable.rb]: the base class uses an integer and the sub-classes change it to an array so they can prepend their own priority

Comment (grosser) [reply plugins/env/app/models/environment_variable.rb]: hmmm yeah I'll change the other one to array too

Comment (adammw) [reply plugins/env/app/models/environment_variable.rb]: can you put that in a comment somewhere too...

File: lib/samson/hooks.rb (hunk 3 above)
Comment (adammw) [reply lib/samson/hooks.rb]: this is where a strongly typed language would help :p

File: plugins/kubernetes/test/models/kubernetes/usage_limit_test.rb [hunk 6]
```diff
@@ -79,4 +79,17 @@ def create_limit(attributes = {})
       Kubernetes::UsageLimit.most_specific(project, deploy_group).must_equal found
     end
   end
+
+  describe "#priority" do
+    let(:limit) { create_limit(scope: environments(:staging)) }
+
+    it "is high with project" do
```
Comment (adammw) [inline plugins/kubernetes/test/models/kubernetes/usage_limit_test.rb]: aren't there other types to check too?

File: plugins/kubernetes/app/models/kubernetes/template_filler.rb (hunk 4 above)
Comment (adammw) [reply plugins/kubernetes/app/models/kubernetes/template_filler.rb]: would the deploy's deploy group ever not match the stages?

Comment (adammw) [reply plugins/kubernetes/app/models/kubernetes/template_filler.rb]: oh, this is for a single deploy group? that makes more sense...

**Review by adammw — APPROVED** [review]
can't understand everything but looks good on the whole.

File: plugins/kubernetes/test/models/kubernetes/usage_limit_test.rb (hunk 6 above)
Comment (grosser) [reply plugins/kubernetes/test/models/kubernetes/usage_limit_test.rb]: usage limits only have project or are global

**Pull request merged**